compiler/rustc_data_structures/src/mini_map.rs

```diff
@@ -0,0 +1,59 @@
+use crate::fx::FxHashMap;
+use arrayvec::ArrayVec;
+use std::hash::Hash;
+
+/// Small-storage-optimized implementation of a map.
+///
+/// Stores elements in a small array up to a certain length
+/// and switches to `HashMap` when that length is exceeded.
+pub enum MiniMap<K, V> {
+    Array(ArrayVec<[(K, V); 8]>),
```

Comment:
Might be pre-mature generalization, but I think you could make this generic over the length of the array too.
```suggestion
pub enum MiniMap<K, V, const N = 8> {
    Array(ArrayVec<[(K, V); N]>),
```

Reply:
I hade tried but I was not able to get it to work with how arrayvec is implemented.
I get this fault:

    error[E0277]: the trait bound `[T; N]: arrayvec::Array` is not satisfied in `arrayvec::ArrayVec<[T; N]>`
      --> compiler/rustc_data_structures/src/mini_set.rs:10:11
       |
    10 |     Array(ArrayVec<Array<T, N>>),
       |           ^^^^^^^^^^^^^^^^^^^^^ within `arrayvec::ArrayVec<[T; N]>`, the trait `arrayvec::Array` is not implemented for `[T; N]`
       |
       = help: the following implementations were found:
                 <[T; 0] as arrayvec::Array>
                 <[T; 100] as arrayvec::Array>
                 <[T; 1024] as arrayvec::Array>
                 <[T; 10] as arrayvec::Array>
               and 53 others
       = note: required because it appears within the type `arrayvec::ArrayVec<[T; N]>`
       = note: no field of an enum variant may have a dynamically sized type
       = help: change the field's type to have a statically known size
    help: borrowed types always have a statically known size
       |
    10 |     Array(&ArrayVec<Array<T, N>>),
       |           ^
    help: the `Box` type always has a statically known size and allocates its contents in the heap
       |
    10 |     Array(Box<ArrayVec<Array<T, N>>>),
       |           ^^^^                     ^

Reply:
I know it is possible for Set, it just has to be changed to accept single type argument - array, just like arrayvec does.
I'm not sure if it's realistically possible for map, because then you'll need a way to constrain array to be an array of pairs and then some way to extract key and value types from said pair. Not sure if it's possible without an additional bunch of hacky hardcoded traits.

Either way both solutions are quite cumbersome and have no real use as of now.

Reply:
Oh haha it doesn't work because `arrayvec` itself hard coded some impls and didn't use const generics.

That's fine then, no need to add it. I just thought it would be nice.